// @example-angular-liferay$angular%2Fplatform-browser@6.1.10/esm5/src/dom/debug/ng_probe.js
Liferay.Loader.define('@example-angular-liferay$angular/platform-browser@6.1.10/esm5/src/dom/debug/ng_probe', ['module', 'exports', 'require'], function (module, exports, require) {
    var define = undefined;
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    import * as tslib_1 from "tslib";
    import * as core from '@angular/core';
    import { exportNgVar } from '../util';
    var CORE_TOKENS = {
        'ApplicationRef': core.ApplicationRef,
        'NgZone': core.NgZone
    };
    var INSPECT_GLOBAL_NAME = 'probe';
    var CORE_TOKENS_GLOBAL_NAME = 'coreTokens';
    /**
     * Returns a {@link DebugElement} for the given native DOM element, or
     * null if the given native element does not have an Angular view associated
     * with it.
     */
    export function inspectNativeElement(element) {
        return core.getDebugNode(element);
    }
    export function _createNgProbe(coreTokens) {
        exportNgVar(INSPECT_GLOBAL_NAME, inspectNativeElement);
        exportNgVar(CORE_TOKENS_GLOBAL_NAME, tslib_1.__assign({}, CORE_TOKENS, _ngProbeTokensToMap(coreTokens || [])));
        return function () {
            return inspectNativeElement;
        };
    }
    function _ngProbeTokensToMap(tokens) {
        return tokens.reduce(function (prev, t) {
            return prev[t.name] = t.token, prev;
        }, {});
    }
    /**
     * Providers which support debugging Angular applications (e.g. via `ng.probe`).
     */
    export var ELEMENT_PROBE_PROVIDERS = [{
        provide: core.APP_INITIALIZER,
        useFactory: _createNgProbe,
        deps: [[core.NgProbeToken, new core.Optional()]],
        multi: true
    }];
});
