// @example-angular-liferay$angular%2Fplatform-browser@6.1.10/esm5/src/platform-browser.js
Liferay.Loader.define('@example-angular-liferay$angular/platform-browser@6.1.10/esm5/src/platform-browser', ['module', 'exports', 'require'], function (module, exports, require) {
  var define = undefined;
  /**
   * @license
   * Copyright Google Inc. All Rights Reserved.
   *
   * Use of this source code is governed by an MIT-style license that can be
   * found in the LICENSE file at https://angular.io/license
   */
  export { BrowserModule, platformBrowser } from './browser';
  export { Meta } from './browser/meta';
  export { Title } from './browser/title';
  export { disableDebugTools, enableDebugTools } from './browser/tools/tools';
  export { BrowserTransferStateModule, TransferState, makeStateKey } from './browser/transfer_state';
  export { By } from './dom/debug/by';
  export { DOCUMENT } from './dom/dom_tokens';
  export { EVENT_MANAGER_PLUGINS, EventManager } from './dom/events/event_manager';
  export { HAMMER_GESTURE_CONFIG, HAMMER_LOADER, HammerGestureConfig } from './dom/events/hammer_gestures';
  export { DomSanitizer } from './security/dom_sanitization_service';
  export * from './private_export';
  export { VERSION } from './version';
});
