// @example-angular-liferay$angular%2Fcommon@6.1.10/esm5/src/platform_id.js
Liferay.Loader.define('@example-angular-liferay$angular/common@6.1.10/esm5/src/platform_id', ['module', 'exports', 'require'], function (module, exports, require) {
  var define = undefined;
  /**
   * @license
   * Copyright Google Inc. All Rights Reserved.
   *
   * Use of this source code is governed by an MIT-style license that can be
   * found in the LICENSE file at https://angular.io/license
   */
  export var PLATFORM_BROWSER_ID = 'browser';
  export var PLATFORM_SERVER_ID = 'server';
  export var PLATFORM_WORKER_APP_ID = 'browserWorkerApp';
  export var PLATFORM_WORKER_UI_ID = 'browserWorkerUi';
  /**
   * Returns whether a platform id represents a browser platform.
   * @experimental
   */
  export function isPlatformBrowser(platformId) {
    return platformId === PLATFORM_BROWSER_ID;
  }
  /**
   * Returns whether a platform id represents a server platform.
   * @experimental
   */
  export function isPlatformServer(platformId) {
    return platformId === PLATFORM_SERVER_ID;
  }
  /**
   * Returns whether a platform id represents a web worker app platform.
   * @experimental
   */
  export function isPlatformWorkerApp(platformId) {
    return platformId === PLATFORM_WORKER_APP_ID;
  }
  /**
   * Returns whether a platform id represents a web worker UI platform.
   * @experimental
   */
  export function isPlatformWorkerUi(platformId) {
    return platformId === PLATFORM_WORKER_UI_ID;
  }
});
