// @example-angular-liferay$angular%2Frouter@6.1.10/esm5/src/route_reuse_strategy.js
Liferay.Loader.define("@example-angular-liferay$angular/router@6.1.10/esm5/src/route_reuse_strategy", ['module', 'exports', 'require'], function (module, exports, require) {
  var define = undefined;
  /**
   * @license
   * Copyright Google Inc. All Rights Reserved.
   *
   * Use of this source code is governed by an MIT-style license that can be
   * found in the LICENSE file at https://angular.io/license
   */
  /**
   * @description
   *
   * Provides a way to customize when activated routes get reused.
   *
   * @experimental
   */
  var RouteReuseStrategy = /** @class */function () {
    function RouteReuseStrategy() {}
    return RouteReuseStrategy;
  }();
  export { RouteReuseStrategy };
  /**
   * Does not detach any subtrees. Reuses routes as long as their route config is the same.
   */
  var DefaultRouteReuseStrategy = /** @class */function () {
    function DefaultRouteReuseStrategy() {}
    DefaultRouteReuseStrategy.prototype.shouldDetach = function (route) {
      return false;
    };
    DefaultRouteReuseStrategy.prototype.store = function (route, detachedTree) {};
    DefaultRouteReuseStrategy.prototype.shouldAttach = function (route) {
      return false;
    };
    DefaultRouteReuseStrategy.prototype.retrieve = function (route) {
      return null;
    };
    DefaultRouteReuseStrategy.prototype.shouldReuseRoute = function (future, curr) {
      return future.routeConfig === curr.routeConfig;
    };
    return DefaultRouteReuseStrategy;
  }();
  export { DefaultRouteReuseStrategy };
});
