// @example-angular-liferay$angular%2Fcore@6.1.10/esm2015/testing/src/logger.js
Liferay.Loader.define('@example-angular-liferay$angular/core@6.1.10/esm2015/testing/src/logger', ['module', 'exports', 'require'], function (module, exports, require) {
  var define = undefined;
  /**
   * @fileoverview added by tsickle
   * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
   */
  /**
   * @license
   * Copyright Google Inc. All Rights Reserved.
   *
   * Use of this source code is governed by an MIT-style license that can be
   * found in the LICENSE file at https://angular.io/license
   */
  import { Injectable } from '@angular/core';
  export class Log {
    constructor() {
      this.logItems = [];
    }
    /**
     * @param {?} value
     * @return {?}
     */
    add(value /** TODO #9100 */) {
      this.logItems.push(value);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    fn(value /** TODO #9100 */) {
      return (a1 = null, a2 = null, a3 = null, a4 = null, a5 = null) => {
        this.logItems.push(value);
      };
    }
    /**
     * @return {?}
     */
    clear() {
      this.logItems = [];
    }
    /**
     * @return {?}
     */
    result() {
      return this.logItems.join('; ');
    }
  }
  Log.decorators = [{ type: Injectable }];
  /** @nocollapse */
  Log.ctorParameters = () => [];
  if (false) {
    /** @type {?} */
    Log.prototype.logItems;
  }
});
