// @example-angular-liferay$angular%2Fcore@6.1.10/esm2015/testing/src/testing.js
Liferay.Loader.define('@example-angular-liferay$angular/core@6.1.10/esm2015/testing/src/testing', ['module', 'exports', 'require'], function (module, exports, require) {
  var define = undefined;
  /**
   * @fileoverview added by tsickle
   * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
   */
  /**
   * @license
   * Copyright Google Inc. All Rights Reserved.
   *
   * Use of this source code is governed by an MIT-style license that can be
   * found in the LICENSE file at https://angular.io/license
   */
  /**
   * @module
   * @description
   * Entry point for all public APIs of the core/testing package.
   */
  export { async } from './async';
  export { ComponentFixture } from './component_fixture';
  export { resetFakeAsyncZone, fakeAsync, tick, flush, discardPeriodicTasks, flushMicrotasks } from './fake_async';
  export { getTestBed, inject, withModule, TestComponentRenderer, ComponentFixtureAutoDetect, ComponentFixtureNoNgZone, TestBed, InjectSetupWrapper } from './test_bed';
  export { __core_private_testing_placeholder__ } from './before_each';
  export { ɵTestingCompiler, ɵTestingCompilerFactory } from './private_export_testing';
});
