// @example-angular-liferay$angular%2Fcore@6.1.10/esm5/src/change_detection/differs/iterable_differs.js
Liferay.Loader.define('@example-angular-liferay$angular/core@6.1.10/esm5/src/change_detection/differs/iterable_differs', ['module', 'exports', 'require'], function (module, exports, require) {
    var define = undefined;
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    import { defineInjectable } from '../../di/defs';
    import { Optional, SkipSelf } from '../../di/metadata';
    import { DefaultIterableDifferFactory } from '../differs/default_iterable_differ';
    /**
     * A repository of different iterable diffing strategies used by NgFor, NgClass, and others.
     *
     */
    var IterableDiffers = /** @class */function () {
        function IterableDiffers(factories) {
            this.factories = factories;
        }
        IterableDiffers.create = function (factories, parent) {
            if (parent != null) {
                var copied = parent.factories.slice();
                factories = factories.concat(copied);
            }
            return new IterableDiffers(factories);
        };
        /**
         * Takes an array of {@link IterableDifferFactory} and returns a provider used to extend the
         * inherited {@link IterableDiffers} instance with the provided factories and return a new
         * {@link IterableDiffers} instance.
         *
         * @usageNotes
         * ### Example
         *
         * The following example shows how to extend an existing list of factories,
         * which will only be applied to the injector for this component and its children.
         * This step is all that's required to make a new {@link IterableDiffer} available.
         *
         * ```
         * @Component({
         *   viewProviders: [
         *     IterableDiffers.extend([new ImmutableListDiffer()])
         *   ]
         * })
         * ```
         */
        IterableDiffers.extend = function (factories) {
            return {
                provide: IterableDiffers,
                useFactory: function (parent) {
                    if (!parent) {
                        // Typically would occur when calling IterableDiffers.extend inside of dependencies passed
                        // to
                        // bootstrap(), which would override default pipes instead of extending them.
                        throw new Error('Cannot extend IterableDiffers without a parent injector');
                    }
                    return IterableDiffers.create(factories, parent);
                },
                // Dependency technically isn't optional, but we can provide a better error message this way.
                deps: [[IterableDiffers, new SkipSelf(), new Optional()]]
            };
        };
        IterableDiffers.prototype.find = function (iterable) {
            var factory = this.factories.find(function (f) {
                return f.supports(iterable);
            });
            if (factory != null) {
                return factory;
            } else {
                throw new Error("Cannot find a differ supporting object '" + iterable + "' of type '" + getTypeNameForDebugging(iterable) + "'");
            }
        };
        IterableDiffers.ngInjectableDef = defineInjectable({
            providedIn: 'root',
            factory: function () {
                return new IterableDiffers([new DefaultIterableDifferFactory()]);
            }
        });
        return IterableDiffers;
    }();
    export { IterableDiffers };
    export function getTypeNameForDebugging(type) {
        return type['name'] || typeof type;
    }
});
