// @example-angular-liferay$angular%2Fcore@6.1.10/esm2015/testing/public_api.js
Liferay.Loader.define('@example-angular-liferay$angular/core@6.1.10/esm2015/testing/public_api', ['module', 'exports', 'require'], function (module, exports, require) {
  var define = undefined;
  /**
   * @fileoverview added by tsickle
   * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
   */
  /**
   * @license
   * Copyright Google Inc. All Rights Reserved.
   *
   * Use of this source code is governed by an MIT-style license that can be
   * found in the LICENSE file at https://angular.io/license
   */
  /**
   * @module
   * @description
   * Entry point for all public APIs of this package.
   */
  export { async, ComponentFixture, resetFakeAsyncZone, fakeAsync, tick, flush, discardPeriodicTasks, flushMicrotasks, getTestBed, inject, withModule, TestComponentRenderer, ComponentFixtureAutoDetect, ComponentFixtureNoNgZone, TestBed, InjectSetupWrapper, __core_private_testing_placeholder__, ɵTestingCompiler, ɵTestingCompilerFactory } from './src/testing';
  // This file only reexports content of the `src` folder. Keep it that way.
});
