// @example-angular-liferay$angular%2Fanimations@6.1.10/esm5/browser/src/dsl/animation_transition_instruction.js
Liferay.Loader.define("@example-angular-liferay$angular/animations@6.1.10/esm5/browser/src/dsl/animation_transition_instruction", ['module', 'exports', 'require'], function (module, exports, require) {
    var define = undefined;
    export function createTransitionInstruction(element, triggerName, fromState, toState, isRemovalTransition, fromStyles, toStyles, timelines, queriedElements, preStyleProps, postStyleProps, totalTime, errors) {
        return {
            type: 0 /* TransitionAnimation */
            , element: element,
            triggerName: triggerName,
            isRemovalTransition: isRemovalTransition,
            fromState: fromState,
            fromStyles: fromStyles,
            toState: toState,
            toStyles: toStyles,
            timelines: timelines,
            queriedElements: queriedElements,
            preStyleProps: preStyleProps,
            postStyleProps: postStyleProps,
            totalTime: totalTime,
            errors: errors
        };
    }
});
