// @example-angular-liferay$angular%2Fanimations@6.1.10/esm5/browser/src/dsl/animation_trigger.js
Liferay.Loader.define('@example-angular-liferay$angular/animations@6.1.10/esm5/browser/src/dsl/animation_trigger', ['module', 'exports', 'require'], function (module, exports, require) {
    var define = undefined;
    import { AnimationStateStyles, AnimationTransitionFactory } from './animation_transition_factory';
    /**
     * @experimental Animation support is experimental.
     */
    export function buildTrigger(name, ast) {
        return new AnimationTrigger(name, ast);
    }
    /**
    * @experimental Animation support is experimental.
    */
    var AnimationTrigger = /** @class */function () {
        function AnimationTrigger(name, ast) {
            var _this = this;
            this.name = name;
            this.ast = ast;
            this.transitionFactories = [];
            this.states = {};
            ast.states.forEach(function (ast) {
                var defaultParams = ast.options && ast.options.params || {};
                _this.states[ast.name] = new AnimationStateStyles(ast.style, defaultParams);
            });
            balanceProperties(this.states, 'true', '1');
            balanceProperties(this.states, 'false', '0');
            ast.transitions.forEach(function (ast) {
                _this.transitionFactories.push(new AnimationTransitionFactory(name, ast, _this.states));
            });
            this.fallbackTransition = createFallbackTransition(name, this.states);
        }
        Object.defineProperty(AnimationTrigger.prototype, "containsQueries", {
            get: function () {
                return this.ast.queryCount > 0;
            },
            enumerable: true,
            configurable: true
        });
        AnimationTrigger.prototype.matchTransition = function (currentState, nextState, element, params) {
            var entry = this.transitionFactories.find(function (f) {
                return f.match(currentState, nextState, element, params);
            });
            return entry || null;
        };
        AnimationTrigger.prototype.matchStyles = function (currentState, params, errors) {
            return this.fallbackTransition.buildStyles(currentState, params, errors);
        };
        return AnimationTrigger;
    }();
    export { AnimationTrigger };
    function createFallbackTransition(triggerName, states) {
        var matchers = [function (fromState, toState) {
            return true;
        }];
        var animation = { type: 2 /* Sequence */, steps: [], options: null };
        var transition = {
            type: 1 /* Transition */
            , animation: animation,
            matchers: matchers,
            options: null,
            queryCount: 0,
            depCount: 0
        };
        return new AnimationTransitionFactory(triggerName, transition, states);
    }
    function balanceProperties(obj, key1, key2) {
        if (obj.hasOwnProperty(key1)) {
            if (!obj.hasOwnProperty(key2)) {
                obj[key2] = obj[key1];
            }
        } else if (obj.hasOwnProperty(key2)) {
            obj[key1] = obj[key2];
        }
    }
});
