// @example-angular-liferay$angular%2Fcore@6.1.10/esm5/testing/src/logger.js
Liferay.Loader.define('@example-angular-liferay$angular/core@6.1.10/esm5/testing/src/logger', ['module', 'exports', 'require'], function (module, exports, require) {
    var define = undefined;
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    import * as tslib_1 from "tslib";
    import { Injectable } from '@angular/core';
    var Log = /** @class */function () {
        function Log() {
            this.logItems = [];
        }
        Log.prototype.add = function (value /** TODO #9100 */) {
            this.logItems.push(value);
        };
        Log.prototype.fn = function (value /** TODO #9100 */) {
            var _this = this;
            return function (a1, a2, a3, a4, a5) {
                if (a1 === void 0) {
                    a1 = null;
                }
                if (a2 === void 0) {
                    a2 = null;
                }
                if (a3 === void 0) {
                    a3 = null;
                }
                if (a4 === void 0) {
                    a4 = null;
                }
                if (a5 === void 0) {
                    a5 = null;
                }
                _this.logItems.push(value);
            };
        };
        Log.prototype.clear = function () {
            this.logItems = [];
        };
        Log.prototype.result = function () {
            return this.logItems.join('; ');
        };
        Log = tslib_1.__decorate([Injectable(), tslib_1.__metadata("design:paramtypes", [])], Log);
        return Log;
    }();
    export { Log };
});
