// @example-angular-liferay$angular%2Fcompiler@6.1.10/src/ml_parser/xml_tags.js
Liferay.Loader.define("@example-angular-liferay$angular/compiler@6.1.10/src/ml_parser/xml_tags", ['module', 'exports', 'require', '@example-angular-liferay$angular/compiler/src/ml_parser/tags'], function (module, exports, require) {
    var define = undefined;
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    (function (factory) {
        if (typeof module === "object" && typeof module.exports === "object") {
            var v = factory(require, exports);
            if (v !== undefined) module.exports = v;
        } else if (typeof define === "function" && define.amd) {
            define("@angular/compiler/src/ml_parser/xml_tags", ["require", "exports", "@angular/compiler/src/ml_parser/tags"], factory);
        }
    })(function (require, exports) {
        "use strict";

        Object.defineProperty(exports, "__esModule", { value: true });
        var tags_1 = require("@example-angular-liferay$angular/compiler/src/ml_parser/tags");
        var XmlTagDefinition = /** @class */function () {
            function XmlTagDefinition() {
                this.closedByParent = false;
                this.contentType = tags_1.TagContentType.PARSABLE_DATA;
                this.isVoid = false;
                this.ignoreFirstLf = false;
                this.canSelfClose = true;
            }
            XmlTagDefinition.prototype.requireExtraParent = function (currentParent) {
                return false;
            };
            XmlTagDefinition.prototype.isClosedByChild = function (name) {
                return false;
            };
            return XmlTagDefinition;
        }();
        exports.XmlTagDefinition = XmlTagDefinition;
        var _TAG_DEFINITION = new XmlTagDefinition();
        function getXmlTagDefinition(tagName) {
            return _TAG_DEFINITION;
        }
        exports.getXmlTagDefinition = getXmlTagDefinition;
    });
});
