// @example-angular-liferay$angular%2Fanimations@6.1.10/esm2015/src/version.js
Liferay.Loader.define('@example-angular-liferay$angular/animations@6.1.10/esm2015/src/version', ['module', 'exports', 'require'], function (module, exports, require) {
  var define = undefined;
  /**
   * @fileoverview added by tsickle
   * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
   */
  /**
   * @license
   * Copyright Google Inc. All Rights Reserved.
   *
   * Use of this source code is governed by an MIT-style license that can be
   * found in the LICENSE file at https://angular.io/license
   */
  /**
   * @module
   * @description
   * Entry point for all public APIs of the animation package.
   */
  /**
   * \@description
   *
   * Represents the version of angular/animations
   */
  export class Version {
    /**
     * @param {?} full
     */
    constructor(full) {
      this.full = full;
      /** @type {?} */
      const splits = full.split('.');
      this.major = splits[0];
      this.minor = splits[1];
      this.patch = splits.slice(2).join('.');
    }
  }
  if (false) {
    /** @type {?} */
    Version.prototype.major;
    /** @type {?} */
    Version.prototype.minor;
    /** @type {?} */
    Version.prototype.patch;
    /** @type {?} */
    Version.prototype.full;
  }
  /** @type {?} */
  export const VERSION = new Version('6.1.10');
});
