// example-angular-liferay$rxjs@6.4.0/internal/operators/subscribeOn.js
"use strict";

Liferay.Loader.define("example-angular-liferay$rxjs@6.4.0/internal/operators/subscribeOn", ['module', 'exports', 'require', '../observable/SubscribeOnObservable'], function (module, exports, require) {
    var define = undefined;
    Object.defineProperty(exports, "__esModule", { value: true });
    var SubscribeOnObservable_1 = require("../observable/SubscribeOnObservable");
    function subscribeOn(scheduler, delay) {
        if (delay === void 0) {
            delay = 0;
        }
        return function subscribeOnOperatorFunction(source) {
            return source.lift(new SubscribeOnOperator(scheduler, delay));
        };
    }
    exports.subscribeOn = subscribeOn;
    var SubscribeOnOperator = function () {
        function SubscribeOnOperator(scheduler, delay) {
            this.scheduler = scheduler;
            this.delay = delay;
        }
        SubscribeOnOperator.prototype.call = function (subscriber, source) {
            return new SubscribeOnObservable_1.SubscribeOnObservable(source, this.delay, this.scheduler).subscribe(subscriber);
        };
        return SubscribeOnOperator;
    }();
});
