// example-angular-liferay$rxjs@6.4.0/_esm5/internal/operators/subscribeOn.js
Liferay.Loader.define('example-angular-liferay$rxjs@6.4.0/_esm5/internal/operators/subscribeOn', ['module', 'exports', 'require'], function (module, exports, require) {
    var define = undefined;
    /** PURE_IMPORTS_START _observable_SubscribeOnObservable PURE_IMPORTS_END */
    import { SubscribeOnObservable } from '../observable/SubscribeOnObservable';
    export function subscribeOn(scheduler, delay) {
        if (delay === void 0) {
            delay = 0;
        }
        return function subscribeOnOperatorFunction(source) {
            return source.lift(new SubscribeOnOperator(scheduler, delay));
        };
    }
    var SubscribeOnOperator = /*@__PURE__*/function () {
        function SubscribeOnOperator(scheduler, delay) {
            this.scheduler = scheduler;
            this.delay = delay;
        }
        SubscribeOnOperator.prototype.call = function (subscriber, source) {
            return new SubscribeOnObservable(source, this.delay, this.scheduler).subscribe(subscriber);
        };
        return SubscribeOnOperator;
    }();
});
