// @example-angular-liferay$angular%2Fcore@6.1.10/esm5/src/errors.js
Liferay.Loader.define('@example-angular-liferay$angular/core@6.1.10/esm5/src/errors', ['module', 'exports', 'require'], function (module, exports, require) {
    var define = undefined;
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    import * as tslib_1 from "tslib";
    export var ERROR_TYPE = 'ngType';
    export var ERROR_DEBUG_CONTEXT = 'ngDebugContext';
    export var ERROR_ORIGINAL_ERROR = 'ngOriginalError';
    export var ERROR_LOGGER = 'ngErrorLogger';
    export function getType(error) {
        return error[ERROR_TYPE];
    }
    export function getDebugContext(error) {
        return error[ERROR_DEBUG_CONTEXT];
    }
    export function getOriginalError(error) {
        return error[ERROR_ORIGINAL_ERROR];
    }
    export function getErrorLogger(error) {
        return error[ERROR_LOGGER] || defaultErrorLogger;
    }
    function defaultErrorLogger(console) {
        var values = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            values[_i - 1] = arguments[_i];
        }
        console.error.apply(console, tslib_1.__spread(values));
    }
});
