// @example-angular-liferay$angular%2Fcompiler@6.1.10/esm5/src/ml_parser/xml_tags.js
Liferay.Loader.define('@example-angular-liferay$angular/compiler@6.1.10/esm5/src/ml_parser/xml_tags', ['module', 'exports', 'require'], function (module, exports, require) {
    var define = undefined;
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    import { TagContentType } from './tags';
    var XmlTagDefinition = /** @class */function () {
        function XmlTagDefinition() {
            this.closedByParent = false;
            this.contentType = TagContentType.PARSABLE_DATA;
            this.isVoid = false;
            this.ignoreFirstLf = false;
            this.canSelfClose = true;
        }
        XmlTagDefinition.prototype.requireExtraParent = function (currentParent) {
            return false;
        };
        XmlTagDefinition.prototype.isClosedByChild = function (name) {
            return false;
        };
        return XmlTagDefinition;
    }();
    export { XmlTagDefinition };
    var _TAG_DEFINITION = new XmlTagDefinition();
    export function getXmlTagDefinition(tagName) {
        return _TAG_DEFINITION;
    }
});
